### PR TITLE
feat: WaitInvoice and LN Client NG CLI tests

### PR DESCRIPTION
**Pull request by m1sterc001guy**
`LnInvoice` in Client NG is currently blocking. It will wait for the invoice to be paid before returning. This is difficult for CLI testing, since it requires a separate thread to pay the invoice. The old client fixed this by having two CLI commands `LnInvoice` and `WaitInvoice`. `LnInvoice` would only generate the payment request and `WaitInvoice` would actually block on the `IncomingContract` to receive the ecash notes.

This PR makes a similar change to the client NG. It adds a new flag `wait_for_payment` that allows `LnInvoice` to either be blocking or non-blocking. `WaitInvoice` will now generate the state machine to wait for the `IncomingContract` to be paid.

This PR also adds lightning client NG tests for sending and receiving from the CLN and LND gateway.

### CODE REVIEW

File: devimint/src/main.rs [hunk 1]
```diff
@@ -605,6 +606,194 @@ async fn cli_tests(dev_fed: DevFed) -> Result<()> {
         .unwrap();
     assert_eq!(client_ng_reissue_amt, reissue_amount);
 
+    // OUTGOING: fedimint-cli NG pays LND via CLN gateway
```
Comment (m1sterc001guy) [inline devimint/src/main.rs:609]: This test is getting quite long now, with most of the code between client and client-ng duplicated. Is it time to delete the old client test code? :)

**Review by elsirion — CHANGES_REQUESTED** [review]
This PR would break the guarantees the state machine abstraction gives us. There probably is a better way to go about it.

File: fedimint-cli/src/ng.rs [hunk 2]
```diff
@@ -103,29 +112,24 @@ pub async fn handle_ng_command<D: IDatabase>(
...
 
             let (operation_id, _) = client
-                .create_bolt11_invoice_and_receive(amount, description, expiry_time, active_gateway)
+                .create_bolt11_invoice_and_receive(
+                    amount,
+                    description,
+                    expiry_time,
+                    active_gateway,
+                    wait_for_payment,
+                )
```
Comment (elsirion) [inline fedimint-cli/src/ng.rs:126]: * Rename `create_bolt11_invoice_and_receive` to `create_bolt11_invoice`
* `create_bolt11_invoice` always waits for offer acceptance or rejection internally
* remove `wait_for_payment` argument, never wait for payment

**Review by elsirion — COMMENTED** [review]
I think these two things is all you need to do to achieve non-blockingness, undoing all other changes.

File: modules/fedimint-ln-client/src/lib.rs [hunk 3]
```diff
@@ -275,20 +280,64 @@ impl LightningClientExt for Client {
...
+        let dyn_states = states
+            .into_iter()
+            .map(|s| s.into_dyn(ln_client_id))
+            .collect();
+
+        self.add_state_machines(&mut dbtx, dyn_states).await?;
+
+        dbtx.commit_tx().await;
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply modules/fedimint-ln-client/src/lib.rs]: Removed

File: fedimint-cli/src/ng.rs (hunk 2 above)
Comment (m1sterc001guy) [reply fedimint-cli/src/ng.rs]: Done

File: modules/fedimint-ln-client/src/lib.rs [hunk 4]
```diff
@@ -754,19 +806,9 @@ impl LightningClientModule {
...
-                        offer_txid: txid,
-                        invoice: sm_invoice.clone(),
-                        payment_keypair,
-                    }),
-                },
-            )]
-        });
+        // The generation of the state machine to receive the ecash notes will happen
+        // later depending on input from the command line
+        let sm_gen = Arc::new(move |_txid: TransactionId, _input_idx: u64| vec![]);
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply modules/fedimint-ln-client/src/lib.rs]: Removed

**Comment by m1sterc001guy** [comment]
Outgoing LN tests are flakey. There seems to be an issue when the mint module can't make exact change, the payment fails for some reason. Still debugging.

**Comment by m1sterc001guy** [comment]
Converting to draft until the issues with outgoing payments are fixed (will submit new PR)

**Comment by m1sterc001guy** [comment]
> Converting to draft until the issues with outgoing payments are fixed (will submit new PR)

Would like to merge this first
https://github.com/fedimint/fedimint/pull/2486

File: devimint/src/federation.rs [hunk 5]
```diff
@@ -192,7 +192,8 @@ impl Federation {
                 ))
             }
         };
-        cmd!(self, "switch-gateway", pub_key).run().await?;
+        cmd!(self, "switch-gateway", pub_key.clone()).run().await?;
+        cmd!(self, "ng", "switch-gateway", pub_key).run().await?;
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply devimint/src/federation.rs]: Hmm we'd probably need additional information to assert anything useful here. Will keep this in mind though if we run into any issues

File: fedimint-cli/src/ng.rs [hunk 6]
```diff
@@ -188,6 +192,45 @@ pub async fn handle_ng_command<D: IDatabase>(
...
+                        operation_id: operation_id.to_hex(),
+                        invoice,
+                    })
+                    .unwrap());
+                }
+            }
+            LnReceiveState::Claimed { txid } => {
+                client.await_claim_notes(operation_id, txid).await?;
+                return get_note_summary(&client).await;
+            }
```
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply fedimint-cli/src/ng.rs]: Would adding a `await_primary_module_output` output similar to this issue be better? https://github.com/fedimint/fedimint/issues/2495

**Comment by jkitman** [comment]
This may be easier to test/debug in the new test framework.

File: fedimint-cli/src/ng.rs [hunk 7]
```diff
@@ -236,6 +259,10 @@ struct InfoResponse {
     denominations_msat: TieredSummary,
 }
 
+pub fn parse_operation_id(s: &str) -> anyhow::Result<OperationId> {
+    OperationId::from_hex(s).map_err(Into::into)
+}
+
```
Comment (elsirion) [inline fedimint-cli/src/ng.rs:267]: Made redundant by 65f07cc44b4f150001740e90e99af33bf9b13d0c in #2416. Should I make it a separate PR or can we merge that first and then you don't need this fn anymore?

File: modules/fedimint-ln-client/src/lib.rs [hunk 8]
```diff
@@ -236,18 +239,14 @@ impl LightningClientExt for Client {
...
-        .await;
-        dbtx.commit_tx().await;
+        // Wait for the transaction to be accepted by the federation, otherwise the
+        // invoice will not be able to be paid
+        //self.transaction_updates(operation_id)
+        //    .await
+        //    .await_tx_accepted(txid)
+        //    .await
+        //    .map_err(|e| anyhow::anyhow!("Offer transaction was not accepted: {:?}",
+        // e))?;
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs]: If you comment that in again you can get rid of the while loop in `ClientNg::LnInvoice` as mentioned above.

File: modules/fedimint-ln-client/src/lib.rs [hunk 9]
```diff
@@ -241,12 +241,11 @@ impl LightningClientExt for Client {
...
-        //    .await
-        //    .await_tx_accepted(txid)
-        //    .await
-        //    .map_err(|e| anyhow::anyhow!("Offer transaction was not accepted: {:?}",
-        // e))?;
+        self.transaction_updates(operation_id)
+            .await
+            .await_tx_accepted(txid)
+            .await
+            .map_err(|e| anyhow::anyhow!("Offer transaction was not accepted: {:?}", e))?;
```
Comment (elsirion) [inline modules/fedimint-ln-client/src/lib.rs:248]: Since you commented it in in the second commit you should be able to remove the while loop in the CLI client

File: modules/fedimint-ln-client/src/lib.rs (hunk 8 above)
Comment (m1sterc001guy) [reply modules/fedimint-ln-client/src/lib.rs]: Oops! Apologies for that, meant to do the fix you mentioned :)

File: fedimint-cli/src/ng.rs (hunk 7 above)
Comment (m1sterc001guy) [reply fedimint-cli/src/ng.rs:267]: I'll review that PR and let's merge that first

File: modules/fedimint-ln-client/src/lib.rs (hunk 9 above)
Comment (m1sterc001guy) [reply modules/fedimint-ln-client/src/lib.rs]: Done

File: fedimint-cli/src/ng.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (m1sterc001guy) [reply fedimint-cli/src/ng.rs]: Going to fix this in https://github.com/fedimint/fedimint/pull/2502

File: fedimint-cli/src/ng.rs (hunk 7 above)
Comment (m1sterc001guy) [reply fedimint-cli/src/ng.rs]: Rebased and removed this function

**Review by jkitman — APPROVED** [review]
Approving, so we can continue iterating with at least one working test.  We can change the API back after the GW client is NG and we can write more integration tests.
